README.md
# JQuery Simple Menu Plugin

Simple Menu Plugin is a utility for quickly building a menu of the site

```js
$('.simple-menu').simpleMenu({
  stickyMenu: true,
  slidingLine: true
})
```

[jqueryscript.net's guide](http://www.jqueryscript.net/menu/Responsive-Sticky-Navigation-Plugin-jQuery-Simple-Menu.html)

## Install

### NPM

```
npm install jquery-simple-menu
```

### YARN

```
yarn add jquery-simple-menu
```

## Webpack

Command | Options
---- | ------
start | webpack-dev-server --env development
build | webpack --env production
serv | static build

### Webpack's examples

```
npm run start
npm run build
npm run serv
```

```
yarn run start
yarn run build
yarn run serv
```

## Examples
![Alt Text](https://github.com/ikloster03/jquery-simple-menu/raw/master/images/example.gif)


[jquery-simple-menu example](https://ikloster03.github.io/jquery-simple-menu/)


## Settings

Option | Type | Default | Description
------ | ---- | ------- | -----------
menuSpeedAnimate | int | 600 | speed animation of the menu
pageNavigationSpeedAnimate | int | 1500 | speed animation of the page's navigation 
btnClassMenu | string | 'btn-menu' | css class of the menu's button
stickyMenu | boolean | false | use / don't use sticky menu
stickyMenuClassName | string | 'fixed' | css class of the sticky menu
slidingLine | boolean | false | use / don't use sliding line
slidingLineClassName | string | 'sliding-line' |  css class of the sliding line
slidingLineClassNameActive | string | 'active' | css class of the active sliding line
slidingLineColor | string | '#ffffff' | color of the sliding line
slidingLineHeight | string | '3px' | height of the sliding line
slidingLineSpeedAnimate | int | 200 | speed animation of the sliding line
winMobWidth | int | 500 | width of the Mobile window
trackedClassName | string | 'tracked' | css class of the waypoint tracked

## Browser Support

- Chrome
- Firefox
- Opera
- Safari
- IE

## Framework Support

- JQuery 1.8+
- Waypoints 4.0+


## Dependencies

> sliding line is implemented using [waypoints](https://github.com/imakewebthings/waypoints)


## TODO

- [x] with gulp
- [x] with eslint
- [x] add Framework Support
- [ ] add Browser Support
- [x] with webpack
- [x] no waypoints version
- [ ] no framework version


## Contact me

- Site: [ikloster03.github.io](https://ikloster03.github.io)
- E-mail: <ikloster@yandex.ru>
- Telegram channel: [t.me/ikloster03](https://t.me/ikloster03)

## [CHANGELOG](https://github.com/ikloster03/jquery-simple-menu/blob/master/CHANGELOG.md)


## [LICENSE](https://github.com/ikloster03/jquery-simple-menu/blob/master/LICENSE)

Copyright (c) 2017 Monastyrev Ivan <ikloster@yandex.ru>. Licensed under the [MIT license](https://github.com/ikloster03/jquery-simple-menu/blob/master/LICENSE).


CHANGELOG.md
# Changelog

## 0.6.0

- Add webpack

## 0.5.0

- Add no-waypoints version
- Small refactoring
- Update eslint's rules

## 0.4.0

- Fix mobile bug


## 0.3.0

- Add gulp
- Add eslint
- Add compression plugin
- Add base check input options


## 0.2.7

- First stable version 



gulpfile.js
const gulp = require('gulp');
const eslint = require('gulp-eslint');
const concat = require('gulp-concat');
const uglify = require('gulp-uglify');
const rename = require('gulp-rename');
const header = require('gulp-header');
const footer = require('gulp-footer');
const tap = require('gulp-tap');
const merge = require('merge-stream');
const pkg = require('./package.json');
const path = require('path');
const babel = require('gulp-babel');

let jsFiles = ['src/js/**/*.js'];

function fileHeader(title) {
    return [
        '/*!',
        title + ' - ' + pkg.version,
        'Copyright © 2017 Monastyrev Ivan',
        'Licensed under the MIT license.',
        'https://github.com/ikloster03/jquery-simple-menu/blob/master/LICENSE',
        '*/\n'
    ].join('\n');
}

gulp.task('lint', function() {
    return gulp.src(jsFiles).pipe(eslint('.eslintrc')).pipe(eslint.format())
});

gulp.task('build', function() {
    return gulp.src('src/js/jquery.simple-menu.js')
        .pipe(concat('jquery.simple-menu.js', { newLine: ';' }))
        .pipe(header(';'))
        .pipe(header(fileHeader('JQuery Simple Menu Plugin')))
        .pipe(gulp.dest('lib/'))
        .pipe(rename('jquery.simple-menu.min.js'))
        .pipe(uglify({
            preserveComments: 'some'
        }))
        .pipe(gulp.dest('lib/'));
});

gulp.task('build-noframework', function() {
    return gulp.src('src/js/noframework.simple-menu.js')
        .pipe(babel({
            presets: ['env']
        }))
        .pipe(concat('noframework.simple-menu.js', { newLine: ';' }))
        .pipe(header(';'))
        .pipe(header(fileHeader('JQuery Simple Menu Plugin')))
        .pipe(gulp.dest('lib/'))
        .pipe(rename('noframework.simple-menu.min.js'))
        .pipe(uglify({
            preserveComments: 'some'
        }))
        .pipe(gulp.dest('lib/'));
});

gulp.task('watch', function() {
    gulp.watch(jsFiles, ['lint', 'build', 'build-noframework']);
});

gulp.task('default', ['lint', 'build', 'build-noframework', 'watch']);
gulp.task('jquery', ['lint', 'build', 'watch']);
gulp.task('noframework', ['lint', 'build-noframework', 'watch']);


webpack.config.js
const path = require('path');
const merge = require('webpack-merge');
const devserver = require('./webpack/devserver');
const uglifyJS = require('./webpack/js.uglify');
const babel = require('./webpack/babel');


const PATHS = {
    source: path.join(__dirname, 'src'),
    build: path.join(__dirname, 'build')
};
const common = {
    entry: {
        'jquery': PATHS.source + '/js/jquery.simple-menu.js',
        'noframework': PATHS.source + 'js/noframework.simple-menu.js',
    },
    output: {
        path: PATHS.build,
        filename: 'js/[name].js'
    },
    plugins: [

    ],
};

module.exports = function(env) {
    if (env === 'production') {
        return merge([
            common,
            babel(),
            uglifyJS()
        ]);
    }
    if (env === 'development') {

        return merge([
            common,
            devserver(),
            babel(),
        ]);
    }
};

lib/noframework.simple-menu.js
/*!
JQuery Simple Menu Plugin - 0.5.1
Copyright © 2017 Monastyrev Ivan
Licensed under the MIT license.
https://github.com/ikloster03/jquery-simple-menu/blob/master/LICENSE
*/
;'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

(function () {

    'use strict';

    var defaults = {
        menuSpeedAnimate: 600,
        pageNavigationSpeedAnimate: 1500,
        btnClassMenu: 'btn-menu',
        stickyMenu: false,
        stickyMenuClassName: 'fixed',
        slidingLine: false,
        slidingLineClassName: 'sliding-line',
        slidingLineClassNameActive: 'active',
        slidingLineColor: '#ffffff',
        slidingLineHeight: '3px',
        slidingLineSpeedAnimate: 200,
        winMobWidth: 500,
        trackedClassName: 'tracked'
    };

    var Tools = function () {
        function Tools() {
            _classCallCheck(this, Tools);
        }

        _createClass(Tools, null, [{
            key: 'getMaxOfArray',
            value: function getMaxOfArray(arr) {

                return Math.max.apply(null, arr);
            }
        }, {
            key: 'getKeyByValue',
            value: function getKeyByValue(obj, val) {

                for (var prop in obj) {
                    if (obj.hasOwnProperty(prop)) {
                        if (obj[prop] === val) {
                            return prop;
                        }
                    }
                }
            }
        }]);

        return Tools;
    }();

    var SimpleMenu = function () {
        function SimpleMenu(options) {
            _classCallCheck(this, SimpleMenu);

            options = options || {};
            this._config = Object.assign({}, defaults, options);

            this._prepareClasses();
            this._initSimpleMenu();
            this._prepareSimpleMenuStyles();

            var conf = this._config;
            conf.stickyMenu && this._initStickyMenu();
            conf.slidingLine && this._initSlidingLine();
        }

        _createClass(SimpleMenu, [{
            key: '_prepareClasses',
            value: function _prepareClasses() {
                var self = this;
            }
        }, {
            key: '_prepareSimpleMenuStyles',
            value: function _prepareSimpleMenuStyles() {}
        }, {
            key: '_scrollSpy',
            value: function _scrollSpy() {}
        }, {
            key: '_initSimpleMenu',
            value: function _initSimpleMenu() {}
        }, {
            key: '_initStickyMenu',
            value: function _initStickyMenu() {}
        }, {
            key: '_initSlidingLine',
            value: function _initSlidingLine() {}
        }]);

        return SimpleMenu;
    }();
})();

lib/jquery.simple-menu.js
/*!
JQuery Simple Menu Plugin - 0.6.1
Copyright © 2017 Monastyrev Ivan
Licensed under the MIT license.
https://github.com/ikloster03/jquery-simple-menu/blob/master/LICENSE
*/
;( function( $ ) {

  'use strict';

  var defaults = {
    menuSpeedAnimate:             600,
    pageNavigationSpeedAnimate:   1500,
    btnClassMenu:                 'btn-menu',
    stickyMenu:                   false,
    stickyMenuClassName:          'fixed',
    slidingLine:                  false,
    slidingLineClassName:         'sliding-line',
    slidingLineClassNameActive:   'active',
    slidingLineColor:             '#ffffff',
    slidingLineHeight:            '3px',
    slidingLineSpeedAnimate:      200,
    winMobWidth:                  500,
    trackedClassName:             'tracked'
  };


  function SimpleMenu( element, options, Waypoint ) {

    this.config = $.extend( { }, defaults, options );
    this.element = element;
    this.Waypoint = Waypoint;

    this.init( );

  }


  SimpleMenu.prototype.init = function( ) {

    var conf = this.config;
    this.getClasses( );
    this.simpleMenu( );
    this.simpleMenuStyles( );

    conf.stickyMenu && this.stickyMenu( );
    conf.slidingLine && this.slidingLine( );

  };

  SimpleMenu.prototype.tools = {

    getMaxOfArray: function( arr ) {

        return Math.max.apply(null, arr);

    },

    getKeyByValue: function( obj, value ) {

        for ( var prop in obj ) {
            if ( obj.hasOwnProperty( prop ) ) {
                if ( obj[ prop ] === value ) {
                    return prop;
                }
            }
        }

    }

  };


  SimpleMenu.prototype.getClasses = function( ) {

    var conf = this.config;

    function getClass( name ) {
      return '.' + name;
    };

    this.btnClassMenu = getClass( conf.btnClassMenu );
    this.slidingLineClassName = getClass( conf.slidingLineClassName );
    this.slidingLineClassNameActive = getClass( conf.slidingLineClassNameActive );
    this.trackedClassName = getClass( conf.trackedClassName );

  };

  SimpleMenu.prototype.simpleMenuStyles = function( ) {

    var self = this,
      el = this.element,
      _window = $( window );

    el.css({
      'display': 'block',
      'position': 'relative',
      'width': '100%'
    });

    el.find( 'ul a' ).css({
      'cursor': 'pointer',
      'display': 'block',
      'text-decoration': 'none'
    });

    function refreshSimpleMenuStyles( ) {

      if ( _window.width() < self.config.winMobWidth ) {
        el.find( 'ul' ).css({
          'position': 'absolute',
          'width': '100%'
        });
        el.find( 'li' ).css({
          'display': 'block'
        });
        el.find( self.btnClassMenu ).css({
          'display': 'block',
          'text-decoration': 'none'
        });
      } else {
        el.find( 'ul' ).css({
          'width': '100%'
        });
        el.find( 'li' ).css({
          'display': 'inline-block'
        });
        el.find( self.btnClassMenu ).css({
          'display': 'none'
        });
      }

    }

    refreshSimpleMenuStyles( );

    _window.resize( function( ) {

      refreshSimpleMenuStyles( );

    });

  };


  SimpleMenu.prototype.simpleMenu = function( ) {

    var conf = this.config,
        el = this.element,
        _window = $( window ),
        menuUl = el.find( 'ul' ),
        menuLi = el.find( 'li' ),
        btnMenu = el.find( this.btnClassMenu ),
        winMobWidth = conf.winMobWidth,
        top = 0;

    if ( _window.width( ) < winMobWidth ) {
      menuUl.hide( );
    }

    btnMenu.click( function( ) {

      menuUl.slideToggle( conf.menuSpeedAnimate );

    });

    menuLi.on( 'click', 'a', function( e ) {
      e.preventDefault( );

      var id = $( this ).attr( 'href' );
      top = $( id ).offset( ).top - menuLi.outerHeight( true );


      $( 'body, html' ).animate({
        scrollTop: top
      }, conf.pageNavigationSpeedAnimate);

    });

    _window.resize( function( ) {

      if ( _window.width( ) > winMobWidth ) {
        menuUl.removeAttr( 'style' );
      } else {
        menuUl.hide( );
      }

    });

  };


  SimpleMenu.prototype.stickyMenu = function( ) {

    var conf = this.config,
        stickyMenuClassName = conf.stickyMenuClassName,
        el = this.element,
        _window = $( window ),
        menuClone = $( '<div/>' ).insertAfter( el ),
        menuHeight = 0, // height of menu
        menuPosition = 0, // height of menu
        menuClonePosition = 0, // position of  clone menu
        windowPosition = 0; // position of window

    function addStickyStyles( ) {

      el.addClass( stickyMenuClassName );
      el.css({
        'position': 'fixed',
        'top': '0',
        'left': '0',
        'width': '100%'
      });

    }

    function removeStickyStyles( ) {

      el.removeClass( stickyMenuClassName );
      el.css({
        'position': 'relative',
        'top': '',
        'left': '',
        'width': '100%'
      });

    }

    function refresh( ) {

      menuHeight = el.outerHeight( true );
      menuPosition = menuClone.offset().top;

    }

    function refreshClone() {

      menuClonePosition = menuClonePosition === 0 ? menuPosition : menuClone.offset().top;

    }

    refresh( );

    menuClone.css( 'height', menuHeight ).hide( );

    refreshClone( );

    _window.resize( function () {
        refresh();
        refreshClone( );
    } );

    _window.scroll( function( ) {

      windowPosition = _window.scrollTop( );

      if ( windowPosition >= menuClonePosition ) {
        addStickyStyles( );
        menuClone.show( );
      } else {
        removeStickyStyles( );
        menuClone.hide( );
      }

    });

  };

  SimpleMenu.prototype.scrollSpy = function( heightMenu ) {
    var self = this,
        _window = $( window );


    function getCurrentSection( obj, num ) {
        var arr = [ ];
        for ( var key in obj ) {
          if ( obj.hasOwnProperty( key ) ) {
            var item = obj[ key ];
            if ( item <= num ) {
              arr.push( item );
            }
          }
        }

        if ( arr.length === 0 ) {
            return null;
        } else  if ( arr.length === 1 ) {
            return  self.tools.getKeyByValue( obj, arr[ 0 ] );
        }

        return self.tools.getKeyByValue( obj, self.tools.getMaxOfArray( arr ) );
    }

      var jSections = $( self.trackedClassName );
      var sections = { };
      jSections.each( function( index, element ) {
          sections[ element.id ] = element.offsetTop;
      });

      var curId, lastId;

      _window.scroll( function ( ) {

        var scrollPosition = _window.scrollTop( );

        curId = getCurrentSection( sections, scrollPosition + heightMenu );
        if ( curId ) {
            if ( lastId !== curId ) {
                lastId = curId;
                self.changeMenuItem( curId );
            }
        }

      });
  }


  SimpleMenu.prototype.slidingLine = function( ) {

    var self = this,
        conf = this.config,
        el = this.element,
        _window = $( window ),
        line,
        active = conf.slidingLineClassNameActive,
        activeLi,
        menuLi = el.find( 'li' ),
        lineWidth,
        liPos;

    function insertLine( ) {

      if ( el.find( self.slidingLineClassName ).length === 0 ) {
        line = $( '<div class="' + conf.slidingLineClassName + '">' ).appendTo( el );
      }

    }

    function refreshPosition( ) {

      activeLi = el.find( 'li.' + active );
      if ( activeLi.length !== 0 ) {
        lineWidth = activeLi.outerWidth( );
        liPos = activeLi.position( ).left;
      }

    }

    function setLine( ) {

      line.css({
        'position': 'absolute',
        'background-color': conf.slidingLineColor,
        'bottom': '0',
        'height': conf.slidingLineHeight
      }).animate({
        'left': liPos,
        'width': lineWidth
      }, conf.slidingLineSpeedAnimate);

    }

    function checkMob( ) {

      _window.width( ) > conf.winMobWidth ? line.show( ) : line.hide( );

    }

    function moveLine( ) {

      refreshPosition( );
      setLine( );

    }

    function refreshLine( ) {

      insertLine( );
      moveLine( );
      checkMob( );

    }

      self.changeMenuItem = function ( hash ) {
      menuLi.removeClass( 'active' );
      menuLi.each(function( ) {
        if ( $( this ).children( 'a' ).attr( 'href' ).slice( 1 ) === hash ) {
          $( this ).addClass( 'active' );
          moveLine( );
        }

      });
    }

    refreshLine( );

    if ( !$.isEmptyObject( self.Waypoint ) ) {
      var waypointTracked = $( self.trackedClassName );

      waypointTracked.waypoint({
        handler: function( ) {
            self.changeMenuItem( this.element.id );
        },
        offset: '30%'
      });
      waypointTracked.waypoint({
        handler: function( ) {
            self.changeMenuItem( this.element.id );
        },
        offset: -menuLi.outerHeight( true )
      });
    } else {
        self.scrollSpy( menuLi.outerHeight( true ) );
    }

    _window.resize( function( ) {

      refreshLine( );

    });

  };


  $.fn.simpleMenu = function( options ) {

    var first = this.first( );
    options = options || { };
    var Waypoint = window.Waypoint || { };
    new SimpleMenu( first, options, Waypoint );
    return first;

  };

} )( jQuery );
